docs/learning/guidelines/release-tags-guidelines.md

```diff
@@ -1,66 +1,205 @@
+- Nested APIs that reside inside another API - e.g., class methods and namespace members.
+- Top-level APIs that do not reside inside another API - e.g., functions and classes not defined inside a namespace.
+
+Nested vs top-level and single-package vs cross-package are two orthogonal API categories.
+
+As of iTwin.js 4.8, we are making efforts to reduce and eventually eliminate the visibility of `@internal` APIs outside of the itwinjs-core repository. Each core package's subdirectory will be organized as follows, using `@itwin/core-backend` as an example:
+
 ```
+  src/
+    core-backend.ts
+    internal/
+      cross-package.ts
+      Symbols.ts
+      <files containing internal top-level APIs>
+    <files containing non-internal top-level APIs>
+```
+
+All top-level internal APIs (whether cross-package or single-package) are defined in source files inside `src/internal/` or subdirectories thereof.
+
+All non-internal top-level APIs are defined in source files inside `src/` or subdirectories thereof.
+
+`src/core-backend.ts` is the "barrel" file that defines the package's public API. It should not export any APIs from inside the `src/internal` directory.
+
+`src/internal/cross-package.ts` exports individual top-level cross-package internal APIs from `src/internal/`.
+
+`src/internal/Symbols.ts` defines [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)s that identify nested internal APIs.
+
+When evaluating an existing or new `@internal` API, apply the following policies in order by priority.
 
-*Extensions* API tag should be on its own line after the supported release tag:
+1. If the API does not need to be `@internal` - i.e., it does not meet the criteria defined [above](#choosing-appropriate-release-tags) - apply the appropriate release tag instead
+1. Convert a nested API to a top-level API, if possible (e.g., a static class method that does not access private members of the class can be converted to a top-level function).
+1. Move a top-level API into `src/internal/`
+1. A top-level, cross-package API should be exported from `src/internal/cross-package.ts`. Don't export single-package APIs from `cross-package.ts`- they can be imported wthin the package using a relative path.
+1. For a nested API that cannot be converted to a top-level API, convert its name to a Symbol using the example below as a guide.
+
+### Converting a nested API
+
+Imagine we have the following `@internal` API that we want to make inaccessible to code outside of itwinjs-core:
 
 ```ts
-/** Sample description of an exported API item.
- * @public
- * @extensions
- */
+/** @public */
+export class Thing {
+  /** @internal */
+  public close(): void {
+    // implementation goes here
+  }
+}
+
 ```
+
+Add to `src/internal/Symbols.ts` the following, if a `_close` symbol doesn't already exist (keeping the list of symbols sorted alphabetically):
+
+```ts
+export const _close = Symbol.for("close_INTERNAL_ONLY_DO_NOT_USE");
```

Comment:
Should the name be more specific about the API, e.g. include class or package name? It seems we can easily get symbol name collisions if we only use method name + constant suffix.

Reply:
What problem do you see arising from, say, core-backend and core-common both having a `_close` Symbol with the same description? They will point to the exact same object.

You would never do something like:
```ts
import { _close as closeBackend } from "@itwin/core-backend";
import { _close as closeCommon } from "@itwin/core-common";

export interface Thing {
  [_closeBackend]: string;
  [_closeCommon]: string;
}
```

That would be a compiler error just as trying to define two properties named "close" would.

Reply:
Ok I [take that back](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBBCeBbGBeGBlJAjEAbAdAGYgBOAFAEQAWAprriBQJQDcAUKJLAogEyoZseIqUq16jVmzYBLMFBolCAQ2A0YAFSpyA5jADebGDADaPALoAuOFBK72xs0l5WbdsDoemL16O89sAL7SnNAwUNoe1lq6AoaOPjAUPBQANEbezq7JzmkZAPT5mYjZACYArgAONHnBHOAQeDT4DDpkEbpOJVKhTS0gbR0eXS5SQA).

Reply:
Made this change, resolve if satisfied.